packages/vite/src/node/constants.ts

```diff
@@ -46,6 +46,27 @@ export const DEFAULT_MAIN_FIELDS = [
   'jsnext',
 ]
 
+// using " and $ as conditions would usually avoid using $ or "
+// as it makes difficult to use the condition in command line or json
+// also avoiding a symbol so that dual package hazard won't be a problem
+type SpecialCondition<T extends string> = `vite$"${T}"` & { __brand: never }
+
+/**
+ * A special condition that would be replaced with production or development
+ * depending on NODE_ENV env variable
+ */
+export const DEV_PROD_CONDITION =
+  `vite$"dev-prod"` as SpecialCondition<'dev-prod'>
+
+export const DEFAULT_CONDITIONS = [
+  'module',
+  'browser',
+  'node',
+  DEV_PROD_CONDITION,
+]
```

Comment:
In the meeting I suggested that maybe we should keep using `"production"` and `"development"` still and avoid a magic string. If one of them is added, the other must exist or it'll warn (we'll auto add it anyways). That way both conditions will always exist. Perhaps this is a simpler approach so we don't have to introduce a new option?

Reply:
Don't you think that a lot of people will forget to add `, "production", "development"`? If 99% of the time we need them there, I think `resolve.conditionDevProd: boolean` would lead to a better experience

Reply:
I don't know if it'll be a common thing for people to do. IIUC we have the condition by default, so those who change the conditions are probably plugin/env authors so they'll be able to address this early on. And if we emit a warning, it should nudge the authors to the right direction without too much hassle I feel like 🤔

Reply:
You can't emit a warning if people don't add `production+development` as that would be fine. Oh, you are saying that we can emit a warning if only one is added. What I'm worried about is that a lot of times these won't end up added. I prefer your idea to the custom symbol though, but I would still prefer the boolean. Curious to see what others think.

Reply:
I'm fine with changing the string value to something that isn't that strange to write it directly in the config as long as it indicates it's a special value (It still works now when you write `vite$"dev-prod"` in the config directly instead of importing the variable). To raise a few ideas: `@vite-dev-prod` / `_vite-dev-prod` / `vite:dev-prod` / `@vite:dev-prod`.

> `vite-dev-prod` will always be a footgun with a lot of users forgetting to add it.

I think this is not special to `vite-dev-prod`. All `module`, `node`, `browser` might be forgot to add by users. I think `configDefaults` would help in this case (at least not to forget updating the code when the default changes in future).

Reply:
What about just `vite_env`, in reference to NODE_ENV?

Reply:
For me, `vite_env` feels like it will be replaced with NODE_ENV which is not the case here.

Reply:
Ok, `vite:dev-prod` sounds ok. Other options `development|production`, `dev|prod`